pkg/p2p/server.go

```diff
@@ -0,0 +1,845 @@
+	}); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// getAckStorage must be called with `acksMapLock` taken.
+func (m *MessageServer) getAck(senderID NodeID, topic Topic) Seq {
+	var senderMap map[Topic]Seq
+	if senderMap = m.acksMap[senderID]; senderMap == nil {
+		senderMap = make(map[Topic]Seq)
+		m.acksMap[senderID] = senderMap
+	}
+	return senderMap[topic]
+}
+
+// setAck must be called with `acksMapLock` taken.
+func (m *MessageServer) setAck(senderID NodeID, topic Topic, ack Seq) {
+	var senderMap map[Topic]Seq
+	if senderMap = m.acksMap[senderID]; senderMap == nil {
+		senderMap = make(map[Topic]Seq)
+		m.acksMap[senderID] = senderMap
+	}
+	senderMap[topic] = ack
+}
+
+func (m *MessageServer) verifyStreamMeta(streamMeta *p2p.StreamMeta) error {
+	if streamMeta == nil {
+		return cerror.ErrPeerMessageIllegalMeta.GenWithStackByArgs()
+	}
+
+	if streamMeta.ReceiverId != m.serverID {
+		return cerror.ErrPeerMessageReceiverMismatch.GenWithStackByArgs(
+			m.serverID,            // expected
+			streamMeta.ReceiverId, // actual
+		)
+	}
+
+	if version.ReleaseSemver() == "" {
+		// skip checking versions for some unit test cases
+		return nil
+	}
+
+	clientVer, err := semver.NewVersion(streamMeta.ClientVersion)
+	if err != nil {
+		log.Error("MessageServer: semver failed to parse",
+			zap.String("ver", streamMeta.ClientVersion),
+			zap.Error(err))
+		return cerror.ErrPeerMessageIllegalClientVersion.GenWithStackByArgs(streamMeta.ClientVersion)
+	}
+
+	serverVer := semver.New(version.ReleaseSemver())
+
+	// Only allow clients with the same Major and Minor.
+	if serverVer.Major != clientVer.Major || serverVer.Minor != clientVer.Minor {
+		return cerror.ErrVersionIncompatible.GenWithStackByArgs(version.ReleaseSemver())
+	}
+
+	return nil
+}
+
+type topicSenderPair struct {
+	Topic    string
+	SenderID string
+}
+
+type pendingMessageEntry struct {
+	StreamMeta *p2p.StreamMeta
+	Entry      *p2p.MessageEntry
+}
+
+type streamHandle struct {
+	mu       sync.RWMutex
+	isClosed bool
+	sendCh   chan p2p.SendMessageResponse
+	closeCh  chan struct{}
+
+	streamMeta *p2p.StreamMeta
+}
+
+func (m *MessageServer) newStreamHandle(meta *p2p.StreamMeta) *streamHandle {
+	return &streamHandle{
+		sendCh:     make(chan p2p.SendMessageResponse, m.config.SendChannelSize),
+		closeCh:    make(chan struct{}),
+		streamMeta: meta,
+	}
+}
+
+func (s *streamHandle) Send(ctx context.Context, response p2p.SendMessageResponse) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if s.isClosed {
+		return cerror.ErrPeerMessageInternalSenderClosed.GenWithStackByArgs()
+	}
+
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case s.sendCh <- response:
+	case <-s.closeCh:
+		return cerror.ErrPeerMessageInternalSenderClosed.GenWithStackByArgs()
+	}
+
+	return nil
+}
+
+func (s *streamHandle) Close() {
+	close(s.closeCh)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.isClosed {
+		return
+	}
+	s.isClosed = true
+	close(s.sendCh)
```

Comment:
Maybe you should close(s.closeCh) after acquiring lock.

Reply:
I think doing so may create a deadlock problem, because a sender might be blocked at writing to `sendCh` with `RLock` taken. Closing `closeCh` before taking the lock can force selects blocking on `sendCh` to break.